File: app/src/prod/java/com/mmartin/ghibliapi/data/FilmsRemoteDataSource.java
package com.mmartin.ghibliapi.data;

import android.support.annotation.NonNull;

import com.mmartin.ghibliapi.data.model.Film;
import com.mmartin.ghibliapi.network.GhibliApi;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;

/**
 * Concrete implementation for getting data from the Ghibli API
 * <p>
 * Created by mmartin on 9/12/17.
 */
public class FilmsRemoteDataSource extends FilmsDataSource {
    GhibliApi api;

    @Inject
    public FilmsRemoteDataSource(@NonNull GhibliApi api) {
        this.api = api;
    }

    @Override
    public Observable<List<Film>> getFilms() {
        return api.getFilmService().getFilms();
    }

    @Override
    public Observable<Film> getFilm(@NonNull String id) {
        return api.getFilmService().getFilmById(id);
    }
}

File: app/src/prod/java/com/mmartin/ghibliapi/data/FilmsRemoteDataSource.kt
package com.mmartin.ghibliapi.data

import com.mmartin.ghibliapi.data.model.Film
import com.mmartin.ghibliapi.network.GhibliApi
import io.reactivex.Flowable
import io.reactivex.Single
import javax.inject.Inject

/**
 * Concrete implementation for getting film data from the Ghibli API
 *
 * Created by mmartin on 9/12/17.
 */
class FilmsRemoteDataSource @Inject
constructor(private val api: GhibliApi) : DataSource<Film>() {

    override val allItems: Flowable<List<Film>>
        get() = api.filmService.films

    override fun getItem(id: String): Flowable<Film> {
        return api.filmService.getFilmById(id)
    }
}
